test/expect-tests/dune_rpc_e2e/dune_rpc_subscription.ml

```diff
@@ -0,0 +1,73 @@
+open Stdune
+open Fiber.O
+open Dune_rpc_e2e
+module Dune_rpc = Dune_rpc_private
+module Sub = Dune_rpc.Sub
+module Client = Dune_rpc_impl.Client
+
+let rec await_progress (poll : Dune_rpc.Progress.t Client.Stream.t) =
+  let* res = Client.Stream.next poll in
+  match res with
+  | None ->
+    printfn "client: progress stream finished";
+    Fiber.return ()
+  | Some Success ->
+    printfn "Succeeded";
+    Fiber.return ()
+  | Some Failed ->
+    printfn "Failed";
+    Fiber.return ()
+  | Some (Waiting | In_progress _ | Interrupted) -> await_progress poll
+
+let display_progress : Dune_rpc.Progress.t -> unit = function
+  | Waiting -> printfn "(Waiting)"
+  | In_progress { complete; remaining } ->
+    printfn "(In_progress ((complete %d) (remaining %d)))" complete remaining
+  | Failed -> printfn "(Failed)"
+  | Interrupted -> printfn "(Interrupted)"
+  | Success -> printfn "(Success)"
+
+let%expect_test "cancelling subscription while request is in-flight" =
+  let exec _pid =
+    let pool = Fiber.Pool.create () in
+    Fiber.fork_and_join_unit
+      (fun () -> Fiber.Pool.run pool)
+      (fun () ->
+        run_client (fun client ->
+            files
+              [ ("dune", "(executable (name foo))"); ("foo.ml", "let () = ()") ];
+            let* poll = poll_exn client Dune_rpc.Public.Sub.progress in
+            let* () = dune_build client "./foo.exe" in
+            (* Clear the pending queue *)
+            let* () = await_progress poll in
+            [%expect
+              {|
+              Building ./foo.exe
+              Build ./foo.exe succeeded
+              Succeeded |}];
+            let result = Fiber.Ivar.create () in
+            (* TODO: Putting the cancellation in the pool is probably flaky. The
+               point of this test is to submit the [next] request, then cancel
+               the poll before a response is received. Experimentally, this is
+               the only combination and ordering I could find that does this
+               correctly. *)
+            let* () =
+              Fiber.Pool.task pool ~f:(fun () -> Client.Stream.cancel poll)
+            in
+            let* () =
+              (* Because we've already received all pending progress updates,
```

Comment:
Yeah this test definitely looks fishy to me. I'm also drawing blank on how to make it more reliable.

Is it not possible to test this with a unit test? I think it should be possible to reliably demonstrate the bug over there.

Reply:
In particular, it should be possible in the unit test to avoid the flakiness by co-ordinating things with an ivar. You need to do something like:

```
Fiber.fork_and_join
   (fun () -> Fiber.never (* we never respond to the client's long polling request *))
   (fun () -> Fiber.Ivar.fill ready_to_cancel () (* client should read this ivar and only then cancel *))
|> Fiber.map ~f:fst
```

Reply:
I tried something like this in a unit test, but ran into an issue in which the Stream writer was getting written to too many times. I'll try again.